main.go

```diff
@@ -66,39 +68,147 @@ func queryAllGroups(r *dns.Msg) {
+	handleRequest(w, r, true)
+}
+
+//attempts to truncate the message
+//very conservatively ONLY truncate the extra
+//todo: once https://github.com/miekg/dns/issues/223 is done, use that
+func truncate(m *dns.Msg, limit int) {
+	//we definitely want to enable compression as our first truncation step
+	m.Compress = true
+	if m.IsTsig() != nil {
+		return
+	}
+	if (m.Len() > limit) {
+		var opt = m.IsEdns0()
+		//if there's only 1 thing in extra, and its OPT, don't do anything
+		if len(m.Extra) > 1 || opt == nil {
+			m.Extra = m.Extra[:0]
+			if opt != nil {
+				m.Extra = append(m.Extra, opt)
+			}
+		}
+	}
+	return
+}
+
+func sendFormatError(w dns.ResponseWriter, r *dns.Msg) {
+	m := new(dns.Msg)
+	m.SetRcode(r, dns.RcodeFormatError)
+	w.WriteMsg(m) //ignore write error
+	return
+}
+
+func removeEdns0(m *dns.Msg) {
+	for i, e := range m.Extra {
+		if e.Header().Rrtype == dns.TypeOPT {
+			m.Extra = append(m.Extra[:i], m.Extra[i + 1:]...)
+			break
+		}
+	}
+}
+
+func handleRequest(w dns.ResponseWriter, r *dns.Msg, tcp bool) {
 	kv := llog.KV{"question": "", "type": ""}
-	var m *dns.Msg
-	if len(r.Question) > 0 {
-		kv["type"], _ = dns.TypeToString[r.Question[0].Qtype]
-		kv["question"] = r.Question[0].Name
-		llog.Info("handling request", kv)
-		rr := NewReq{r, make(chan *dns.Msg)}
-		newCh <- rr
-		m = <-rr.ReplyCh
+	var ok bool
+	kv["type"], ok = dns.TypeToString[r.Question[0].Qtype]
+	if !ok || kv["type"] == "" || kv["type"] == "None" {
+		kv["qtype"] = r.Question[0].Qtype
+		llog.Warn("invalid type received", kv)
+		dns.HandleFailed(w, r)
+		return
+	}
+	if len(r.Question) == 0 {
+		llog.Warn("received request with no questions", kv)
+		sendFormatError(w, r)
+		return
+	}
+	kv["question"] = r.Question[0].Name
+
+	//RFC 1035 defines this as the max
+	//though the unfortunate thing is that RFC 5625 says we SHOULD NOT
+	//truncate if the packet is larger than 512 anymore
+	limit := dns.MinMsgSize
+
+	addLimit := true
+	opt := r.IsEdns0()
+	if opt != nil {
+		limit = int(opt.UDPSize())
+		llog.Debug("received Edns0 option")
+		// we don't want to overwrite the existing Edns0 if there is one
+		addLimit = false
+	}
+	if addLimit {
+		//if we don't manually set this to a high number then we might not even
+		//get a response to try and truncate
+		//dig sets its size to 4096 so why can't we?
+		r.SetEdns0(4096, false)
+		llog.Debug("adding Edns0 option")
 	}
+	kv["limit"] = limit
+
+	llog.Info("handling request", kv)
+
+	var m *dns.Msg
+Query:
+	rr := NewReq{r, limit, make(chan *dns.Msg)}
+	newCh <- rr
+	m = <-rr.ReplyCh
 	if m == nil {
-		llog.Info("error handling request", kv)
+		llog.Warn("error handling request", kv)
 		dns.HandleFailed(w, r)
 		return
 	}
+	if addLimit && m.Rcode == dns.RcodeFormatError {
+		//since the server doesn't support Edns0 requests lets remove the one
+		//we added and retry the request
+		addLimit = false
+		removeEdns0(m)
+		llog.Info("received a format error, retrying", kv)
+		goto Query
```

Comment:
Lol i saw it in the go source and was like "ooo that'd work perfectly for me". I can convert it to a loop though